lib/rules/no-extra-parens.js

```diff
@@ -766,6 +766,36 @@ module.exports = {
             return false;
         }
 
+        /**
+         * Checks if the left-hand side of an assignment is an identifier and the right-hand side is
+         * an anonymous class or function.
+         *
+         * As per https://tc39.es/ecma262/#sec-assignment-operators-runtime-semantics-evaluation, an
+         * assignment where the right-hand side is an anonymous class or function and the left-hand
+         * side is an *unparenthesized* identifier has different semantics than other assignments.
+         * Specifically, when an expression like `foo = function () {}` is evaluated, `foo.name`
+         * will be set to the string "foo", i.e. the identifier name. The same thing does not happen
+         * when evaluating `(foo) = function () {}`.
+         * Since the parenthesizing of the identifier in the left-hand side is significant in this
+         * special case, the parentheses, if present, should not be flagged as unnecessary.
+         * @param {ASTNode} node an AssignmentExpression node.
+         * @returns {boolean} `true` if the left-hand side of the assignment is an identifier and
+         * the right-hand side is an anonymous class or function; otherwise, `false`.
+         */
+        function isAnonymousFunctionAssignmentException({ left, right }) {
+            if (left.type === "Identifier") {
+                const rhsType = right.type;
+
+                if (rhsType === "ArrowFunctionExpression") {
+                    return true;
+                }
+                if ((rhsType === "FunctionExpression" || rhsType === "ClassExpression") && !right.id) {
+                    return true;
+                }
+            }
+            return false;
+        }
```

Comment:
Should we also check the operator? Per the spec, this behavior seems to apply only if the operator is one of `=`, `&&=`, `||=`, or `??=`, not e.g. `+=`.

Reply:
@mdjermanovic I think so. I got confused because Babel doesn't seem to care about the operand, and decides to [keep the parentheses](https://babeljs.io/repl/#?browsers=&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=G4QwTgBCDcBQAUICUEDUBeCAzArgOwGMAXASwHs8IAjeFAbwgF85EUNt9jzLaIHnYQA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.20.15&externalPlugins=&assumptions=%7B%7D) even when optimizing `(a) += function () {}`. This made me think that I could be missing some edge case, but actually the spec is clear enough in defining the semantics of each operator (group) in a separate section. I'm going to update the PR like you suggested. Thanks!

Reply:
Updated as per discussion.